### PR TITLE
fix(block-editor): Relationship contentlet crash

### LINKED ISSUE #28626: Block editor field relationship contentlet crash
### Problem Statement
When using a block editor field in dotCMS and adding the same related content to each other, it crashes the contentlet.

### Steps to Reproduce
1. Create a content type with a block editor field.
2. Create 3 pieces of content 1, 2 and 3
3. In the block editor field of 1, relate 2 and 3
4. In the block editor field of 2, relate 1 and 3
5. In the block editor field of 3, relate 1 and 2
6. Save and publish the content
7. The content should be somehow corrupted.

### Acceptance Criteria
Contentlets with block editor fields should be able to relate to each other without causing any crashes.

### External Links
N/A

### Assumptions & Initiation Needs
Looks like we need to save the contentlets in the block editor without their internal related contentlet.

### Quality Assurance Notes & Workarounds
N/A

### LINKED ISSUE #28659: Store relationship references in blockeditor, not the entire related contentlet object
### Parent Issue

_No response_

### Problem Statement

When contentlet is added to the block editor, we are storing the entire related contentlet inside the content block.  This is an anti-pattern.  We need to store the content references in the block editor, not the related content objects.  

- We are seeing nested contentlets included in the blockeditor json
- If the related contentlets are related to other contentlets, the whole graph is being included in the blockeditor json
- If there is a cyclic graph of the related contentlets, we are including them in recursively in the blockeditor json, which is creating huge content objects and blowing memory stacks
- What happens if the related content is unpublished - it is still included in the blockeditor json?  Is it still displayed?

### Steps to Reproduce

1. Create a content type with a block editor field
2. create 3 pieces of content of the type - Content A, Content B, Content C and relate them to each other in the block editor
3. view the resultant json that is stored in the blockeditor - it includes not only the related content, but the content related to the related content, etc....

### Acceptance Criteria

We the content block should only store the identifiers of the related content.

### dotCMS Version

24.04

### Proposed Objective

Reliability

### Proposed Priority

Priority 1 - Show Stopper

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

Postman tests are available in this file: `dotcms-postman/src/main/resources/postman/Content_Resource.postman_collection.json`, under these folders:

- `Contente Resource / Content with Block Editor / Checking Block Editor Recursiveness`
- `Contente Resource / Content with Block Editor / Block Editor and Relationship Fields`

Additionally, take into account that the following tickets have been fixed with the code changes from this ticket. This means that the following must be QA'ed as well:

- https://github.com/dotCMS/core/issues/28664
- https://github.com/dotCMS/core/issues/28631

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by rjvelazco**
### What we did

- Store only the related content's identifiers in the block editor.
- Hydrate the content on Read.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java [hunk 1]
```diff
@@ -34,12 +36,22 @@
...
+    private static final Lazy<Integer> MAX_RECURSION_LEVEL = Lazy.of(() -> Config.getIntProperty("STORY_BLOCK_MAX_RECURSION_LEVEL", 2));
+
     @Override
+    @CloseDBIfOpened
     public StoryBlockReferenceResult refreshReferences(final Contentlet contentlet) {
         final MutableBoolean refreshed = new MutableBoolean(false);
         if (null != contentlet && null != contentlet.getContentType() &&
                 contentlet.getContentType().hasStoryBlockFields()) {
+
+            if (ExceptionUtil.isMethodCallCountGteThan(
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java:48]: don't love this logic - it pulls 1000 lines of the thread stack and concatenates them together for a new String, which is gonna be huge, and then does a find?  

Just loop over the thread stack[] and search for the string in the line and when that number is >= the MAX_RECURSION, then break.  No new Strings allocated.

File: dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java [hunk 2]
```diff
@@ -457,6 +458,19 @@ public static String getCurrentStackTraceAsString (final int limit) {
...
     }
 
+    /**
+     * Checks whether a specific method name with its fully-qualified class name has been called
+     * at least X number of times.
+     * @param methodName
+     * @param maxCount
+     * @return
+     */
+    public static boolean isMethodCallCountGteThan(final String methodName, final int maxCount) {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java:468]: I would expect this method to be in the `ThreadUtil` I think.

**Review by wezell — CHANGES_REQUESTED** [review]
Need to discuss.

File: dotCMS/src/main/java/com/dotcms/content/business/json/ContentletJsonAPIImpl.java [hunk 3]
```diff
@@ -313,7 +314,9 @@ private Map<String, Object> getContentletMapFromImmutable(final Contentlet immut
             if(value instanceof Date){
                 value = new Timestamp(((Date) value).getTime());
             }
-
+            if (field instanceof StoryBlockField) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/business/json/ContentletJsonAPIImpl.java:317]: may add some comment here why we are adding the raw, in a year nobody will remember why we are doing it

File: dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java [hunk 4]
```diff
@@ -133,6 +156,24 @@ private boolean refreshStoryBlockMap(final Map<String, Object> contentMap, final
...
+     * {@link PageMode#ADMIN_MODE}, then the live Inode will be returned. Otherwise, the working
+     * inode is returned.
+     *
+     * @param contentletVersionInfo The {@link ContentletVersionInfo} object containing the
+     *                              Contentlet's Inodes.
+     *
+     * @return The Inode of the Contentlet based on the current Page Mode.
+     */
+    private String getInodeBasedOnPageMode(final ContentletVersionInfo contentletVersionInfo) {
+        final PageMode currentPageMode = PageMode.get(HttpServletRequestThreadLocal.INSTANCE.getRequest());
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java:171]: Nice

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java [hunk 5]
```diff
@@ -34,12 +45,23 @@
...
+    private static final int MAX_RECURSION_LEVEL = 2;
+    private static final Lazy<String> MAX_RELATIONSHIP_DEPTH = Lazy.of(() -> Config.getStringProperty(
+            "STORY_BLOCK_MAX_RELATIONSHIP_DEPTH", "2"));
+
     @Override
+    @CloseDBIfOpened
     public StoryBlockReferenceResult refreshReferences(final Contentlet contentlet) {
         final MutableBoolean refreshed = new MutableBoolean(false);
-        if (null != contentlet && null != contentlet.getContentType() &&
+        boolean inTransaction = DbConnectionFactory.inTransaction();
```
[the remark this answers is not in the document]
Comment (jcastro-dotcms) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java]: Done!

File: dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java [hunk 6]
```diff
@@ -281,15 +288,61 @@ public String toJson (final Map<String, Object> blockEditorMap) throws JsonProce
...
      */
-    private void refreshBlockEditorDataMap(final Map<String, Object> dataMap, final String liveInode) throws DotDataException, DotSecurityException {
-        final Contentlet contentlet = APILocator.getContentletAPI().find(liveInode, APILocator.systemUser(), false);
-        final Set<String> contentFieldNames = dataMap.keySet();
-        for (final String contentFieldName : contentFieldNames) {
-            final Object value = contentlet.get(contentFieldName);
-            if (null != value) {
-                dataMap.put(contentFieldName, value);
+    private void refreshBlockEditorDataMap(Map<String, Object> dataMap, final String liveInode) {
+        final Optional<Contentlet> contentlet = APILocator.getContentletAPI().findInDb(liveInode);
```
[the remark this answers is not in the document]
Comment (jcastro-dotcms) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java]: Done!

File: dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java (hunk 1 above)
Comment (jcastro-dotcms) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/StoryBlockAPIImpl.java]: Done!

File: dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java (hunk 2 above)
Comment (jcastro-dotcms) [reply dotCMS/src/main/java/com/dotcms/exception/ExceptionUtil.java]: Done!

**Review by wezell — APPROVED** [review]
This is looking better.

**Pull request merged**